### PR TITLE
Add pretrained model-family distributed smoke matrix

### PR DESCRIPTION
**Pull request by winglian**
# What does this PR do?
Uses pretrained models across various model families to validate DDP behavior and tests that the loss decreases. 
also includes loss bounds for the model alongside gradient accumulation steps to catch regressions in gradient accumulation losses.

Fixes # (issue)

## Code Agent Policy

The Transformers repo is currently being overwhelmed by a large number of PRs and issue comments written by
code agents. We are currently bottlenecked by our ability to review and respond to them. As a result, 
**we ask that new users do not submit pure code agent PRs** at this time. 
You may use code agents in drafting or to help you diagnose issues. We'd also ask autonomous "OpenClaw"-like agents
not to open any PRs or issues for the moment.

PRs that appear to be fully agent-written will probably be closed without review, and we may block users who do this
repeatedly or maliciously. 

This is a rapidly-evolving situation that's causing significant shockwaves in the open-source community. As a result, 
this policy is likely to be updated regularly in the near future. For more information, please read [`CONTRIBUTING.md`](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md).

- [X] I confirm that this is not a pure code agent PR.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by SunMarc** [comment]
run-slow: trainer

File: tests/trainer/distributed/test_trainer_distributed_ddp.py [hunk 1]
```diff
@@ -34,10 +34,32 @@
...
+    ("qwen2", "axolotl-ai-co/tiny-qwen2-129m", 0.04),
+    ("qwen3", "axolotl-ai-co/tiny-qwen3-129m", 0.04),
+    ("phi", "axolotl-ai-co/tiny-phi-64m", 0.04),
+    ("falcon", "axolotl-ai-co/tiny-falcon-42m", 0.02),
+    ("gemma2", "axolotl-ai-co/tiny-gemma2-137m", 0.29),
+)
+GRADIENT_ACCUMULATION_PARAMS = (1, 4)
+# Per-step tolerance: max(abs_tol, rel_tol * |expected|).
+LOSS_TRAJECTORY_ABS_TOL = 0.05
+LOSS_TRAJECTORY_REL_TOL = 0.10
```
Comment (SunMarc) [inline tests/trainer/distributed/test_trainer_distributed_ddp.py:62]: let's try to minimize the number of global var so that things are cleaner. For example, it should be fine to remove `LOSS_TRAJECTORY_ABS_TOL` or `LOSS_TRAJECTORY_REL_TOL` or even maybe `GRADIENT_ACCUMULATION_PARAMS`. Let's remove as much as possible so that we only keep what we really might need to focus on.

File: tests/trainer/distributed/test_trainer_distributed_ddp.py [hunk 2]
```diff
@@ -47,6 +69,34 @@
...
+    return f"{model_name}:ga={gradient_accumulation_steps}:mode={data_mode}"
+
+
+def _save_loss_trajectory(key, losses):
+    """Write a captured trajectory back into the on-disk reference file."""
+    payload = _load_loss_trajectories()
+    payload.setdefault("trajectories", {})[key] = losses
+    with open(LOSS_TRAJECTORIES_FILE, "w") as f:
+        json.dump(payload, f, indent=2, sort_keys=True)
+        f.write("\n")
```
Comment (SunMarc) [inline tests/trainer/distributed/test_trainer_distributed_ddp.py:99]: there's a lot of things that can be shared across other tests. Maybe we should try to also add it for `fsdp` and `deepspeed` and move most of the stuff to test_trainer_distributed.py ?

**Review by SunMarc — COMMENTED** [review]
Thanks a lot for that ! Maybe we should take this opportunuty to also run the same tests for fsdp and deepspeed and move most of the methods you wrote for ddp to test_trainer_distributed.py ? Btw the tests you added are passing so that quite nice already.
